src/chainstate/stacks/boot/cost-voting.clar

```diff
@@ -0,0 +1,205 @@
+    ((proposal-id uint))
+    ((cost-function-contract principal)
+     (cost-function-name (string-ascii 128))
+     (function-contract principal)
+     (function-name (string-ascii 128))
+     (expiration-block-height uint)))
+
+;; confirmed cost-function proposals
+(define-map confirmed-proposals
+   ((confirmed-id uint))
+   ((confirmed-proposal
+        {  function-contract: principal,
+           function-name: (string-ascii 128),
+           cost-function-contract: principal,
+           cost-function-name: (string-ascii 128),
+           confirmed-height: uint })))
+
+(define-map proposal-confirmed-id ((proposal-id uint)) ((confirmed-id uint)))
+
+(define-map functions-to-confirmed-ids
+   ((function-contract principal) (function-name (string-ascii 50)))
+   ((proposal-id uint)))
+
+;; cost-function proposal votes
+(define-map proposal-votes ((proposal-id uint)) ((votes uint)))
+
+;; cost-function proposal vetos
+(define-map proposal-vetos ((proposal-id uint)) ((vetos uint)))
+
+;; proposal vetos per block
+(define-map exercised-veto ((proposal-id uint) (veto-height uint)) ((vetoed bool)))
+
+;; the number of votes a specific principal has committed to a proposal
+(define-map principal-proposal-votes ((address principal) (proposal-id uint)) ((votes uint)))
+
+;; getter for cost-function proposals
+(define-read-only (get-proposal (proposal-id uint))
+    (map-get? proposals { proposal-id: proposal-id }))
+
+;; getter for confirmed cost-function proposals
+(define-read-only (get-confirmed-proposal (confirmed-id uint))
+    (map-get? confirmed-proposals { confirmed-id: confirmed-id }))
+
+;; getter for cost-function proposal votes
+(define-read-only (get-proposal-votes (proposal-id uint))
+    (get votes (map-get? proposal-votes { proposal-id: proposal-id })))
+
+;; getter for cost-function proposal vetos
+(define-read-only (get-proposal-vetos (proposal-id uint))
+    (get vetos (map-get? proposal-vetos { proposal-id: proposal-id })))
+
+;; getter for cost-function proposal votes, for specific principal
+(define-read-only (get-principal-votes (address principal) (proposal-id uint))
+    (get votes (map-get? principal-proposal-votes { address: address, proposal-id: proposal-id })))
+
+;; Propose cost-functions
+(define-public (submit-proposal (function-contract principal)
+                                (function-name (string-ascii 128))
+                                (cost-function-contract principal)
+                                (cost-function-name (string-ascii 128)))
+    (begin
+        (map-insert proposals { proposal-id: (var-get proposal-count) }
+                              { cost-function-contract: cost-function-contract,
+                                cost-function-name: cost-function-name,
+                                function-contract: function-contract,
+                                function-name: function-name,
+                                expiration-block-height: (+ burn-block-height u2016) })
```

Comment:
I'm curious why `u2016` is a hard-coded value?  Do we want to allow proposals to have variable voting periods (IIRC Bitcoin has allowed voting periods for up to a whole year, for example)?  If not, then `u2016` should be a constant in the code, since you use it in multiple places.